semgrep-core/src/runner/Run_semgrep.ml

```diff
@@ -74,8 +74,37 @@ let replace_named_pipe_by_regular_file path =
 (* for -gen_layer, see Experiments.ml *)
 let _matching_tokens = ref []
 
+let string_of_toks toks =
+  String.concat ", " (Common.map (fun tok -> PI.str_of_info tok) toks)
+
+let rec print_taint_call_trace ~format ~spaces = function
+  | Pattern_match.Toks toks -> Matching_report.print_match ~format ~spaces toks
+  | Call { call_toks; intermediate_toks; call_trace } ->
+      let spaces_string = String.init spaces (fun _ -> ' ') in
+      pr (spaces_string ^ "call to");
+      Matching_report.print_match ~format ~spaces call_toks;
+      if intermediate_toks <> [] then
+        pr
+          (spf "%sthese intermediate values are tainted: %s" spaces_string
+             (string_of_toks intermediate_toks));
+      pr (spaces_string ^ "then");
+      print_taint_call_trace ~format ~spaces:(spaces + 2) call_trace
+
+let print_taint_trace ~format taint_trace =
+  if format <> Matching_report.Normal then ()
+  else
+    let (lazy { Pattern_match.source; tokens; sink }) = taint_trace in
+    pr "  * Taint comes from:";
+    print_taint_call_trace ~format ~spaces:4 source;
+    if tokens <> [] then
+      pr
+        (spf "  * These intermediate values are tainted: %s"
+           (string_of_toks tokens));
+    pr "  * This is how taint reaches the sink:";
```

Comment:
This is needed in DeepSemgrep because we are inter procedural, so the taint may go through several function calls before reaching the sink. But in Semgrep this will never happen so I think it's just redundant.

Reply:
You can still handle it for the "just in case", but instead maybe you can do it conditionally. If `sink` is not a `Call`, then printing it just makes the output more verbose.

Reply:
Since end users don't use semgrep-core directly, this is basically only visible to us so I don't think it's worth spending much time trying to polish this output.

Reply:
I use it a lot so I do care, but I can fix it myself too, so I will open a PR!

Reply:
Sounds good! I'm making a few changes this morning based on feedback here. If you could hold off for a few hours it might help us avoid merge conflicts.

Reply:
If you are working on these comments I guess it's also not a big deal for you to apply this diff?
```diff
@@ -100,8 +100,11 @@ let print_taint_trace ~format taint_trace =
       pr
         (spf "  * These intermediate values are tainted: %s"
            (string_of_toks tokens));
-    pr "  * This is how taint reaches the sink:";
-    print_taint_call_trace ~format ~spaces:4 sink
+    match sink with
+    | Pattern_match.Toks _ -> ()
+    | _call ->
+        pr "  * This is how taint reaches the sink:";
+        print_taint_call_trace ~format ~spaces:4 sink
 
 let print_match ?str match_format mvars mvar_binding ii_of_any
     tokens_matched_code taint_trace =
```
Otherwise I don't think it will cause major conflicts.